pkg/controllers/observability-ui/observability-ui-plugin/controller.go

```diff
@@ -0,0 +1,218 @@
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metaerrors "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	obsui "github.com/rhobs/observability-operator/pkg/apis/observability-ui/v1alpha1"
+	"github.com/rhobs/observability-operator/pkg/reconciler"
+)
+
+type resourceManager struct {
+	k8sClient      client.Client
+	scheme         *runtime.Scheme
+	logger         logr.Logger
+	controller     controller.Controller
+	pluginConf     UIPluginsConfiguration
+	clusterVersion string
+}
+
+type UIPluginsConfiguration struct {
+	Images map[string]string
+}
+
+type Options struct {
+	PluginsConf UIPluginsConfiguration
+}
+
+// RBAC for managing UIPlugins
+// +kubebuilder:rbac:groups=observability.openshift.io,resources=uiplugins,verbs=get;list;watch;create;update;delete;patch
+// +kubebuilder:rbac:groups=observability.openshift.io,resources=uiplugins/status,verbs=get;update
+
+// RBAC for managing observability ui plugin objects
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=serviceaccounts;services;configmaps,verbs=get;list;watch;create;update;patch;delete
+
+// RBAC for managing Console CRs
+// +kubebuilder:rbac:groups=operator.openshift.io,resources=consoles,verbs=get;patch;list;watch
+// +kubebuilder:rbac:groups=console.openshift.io,resources=consoleplugins,verbs=get;list;watch;create;update;delete;patch
+
+// RBAC for reading cluster version
+// +kubebuilder:rbac:groups=config.openshift.io,resources=clusterversions,verbs=get;list;watch
+
+func RegisterWithManager(mgr ctrl.Manager, opts Options) error {
+	logger := ctrl.Log.WithName("observability-ui")
+
+	clusterVersion, err := getClusterVersion(mgr.GetAPIReader())
+
+	if err != nil {
+		logger.Error(err, "failed to get cluster version")
+		return err
+	}
+
+	rm := &resourceManager{
+		k8sClient:      mgr.GetClient(),
+		scheme:         mgr.GetScheme(),
+		logger:         logger,
+		pluginConf:     opts.PluginsConf,
+		clusterVersion: clusterVersion.Status.Desired.Version,
+	}
+
+	generationChanged := builder.WithPredicates(predicate.GenerationChangedPredicate{})
+
+	ctrl, err := ctrl.NewControllerManagedBy(mgr).
+		For(&obsui.UIPlugin{}).
+		Owns(&appsv1.Deployment{}, generationChanged).
+		Owns(&v1.Service{}, generationChanged).
+		Owns(&v1.ServiceAccount{}, generationChanged).
+		Owns(&rbacv1.ClusterRole{}, generationChanged).
+		Owns(&rbacv1.ClusterRoleBinding{}, generationChanged).
+		Owns(&osv1alpha1.ConsolePlugin{}, generationChanged).
```

Comment:
A namespaced resources such as UIPlugin are not allowed to have cluster-scoped depedents. This can lead into poisoning the k8s resource garbage collector like unwanted deletes of the UIPlugin resource. Source: https://kubernetes.io/docs/concepts/architecture/garbage-collection/#owners-dependents

Reply:
I would expect `Owns` to do the right thing here. First and foremost this sets up a watch. And since this controller generates the resource using `Owns` should be the right approach. If this sets an OwnerReference I'd consider this a bug in controller runtime tbh. At least I can't find anything in their documentation that calls this out.

Reply:
I've checked on a live cluster and the controller doesn't seem to set the owner ref for the cluster role. That being said, deleting the UIPlugin object doesn't remove the cluster role (for the same reason).

Reply:
I updated the UIPlugin to be a cluster scoped as it should be, and also changed the ClusterRole and ClusterRoleBinding to Role and RoleBinding. While doing so I realized the reconciler was only considering owner and resources with the same namespace. I wonder if [this](https://github.com/rhobs/observability-operator/pull/434/files#diff-fc2bc6354d4506ced4adfd45dda2737dc917c217e282041d9af5bce57e514a85R29) would be the right way to check for cluster scoped resources that own namespaced ones: